### DISCUSSION TITLE
Optimizer coding problem

### DISCUSSION THREAD

**Discussion by Vfriday** [community/community#73440] [Programming Help]
### Select Topic Area

Question

### Body

I am trying to synthesize H2 observer gains at every sampling time for a varying system. I used interpreted matlab function in simulink using the following code which run too slow:
function L = LMI(ch)
persistent gamma_square alpha beta C
%%%%%%%  Strip Stability Domain Parameters for unit circle %%%%%%%
alpha = -0.75;               % minimum real part of eigenvalues
beta = 0.75;                  % maximum real part of eigenvalues 
A = [];
for i = 1:8:8*8
A = [A,ch(i:i+7)];
end 
% A_aug;
B = [];
for i = 64+1:8:65+15
    B = [B,ch(i:i+7)];
end
C=[eye(2) zeros(2) zeros(2) zeros(2)];
 %%%%%%% LMI Design Using YALMIP %%%%%%%%
yalmip('clear')
% %%% LMI Variables 
X = sdpvar(8,8);                % 8x8 fully parameterized square matrix  
Y = sdpvar(2,8);                % 2x8 matrix   
Z = sdpvar(2,2);                % 2x2 symmetrical matrix 
gamma_square = sdpvar(1,1);     % Upper bound of H2 norm to minimize

% Define your optimization problem with the time-varying constraints
    constraints = [A'*X + X*A + C'*Y + Y'*C + 2*alpha*X <= 0,
                   A'*X + X*A + C'*Y + Y'*C + 2*beta*X >= 0,
                   A'*X + X*A + C'*Y + Y'*C + C'*C <= 0,
                   [-Z, B'*X; X*B, -X] <= 0,
                   trace(Z) <= gamma_square];
% Define the objective function
objective = gamma_square; % Objective function
% Define an optimizer object which solves the problem for a particular
% initial state and reference
ops = sdpsettings('solver','sedumi');   % Choice of solver 
optimize(constraints,objective,ops);
L = -value(X)\value(Y)';           % Observer gain

I tried to speed up the simulation by using optimizer object discussed thus, https://yalmip.github.io/example/simulink/ in order to avoid calling the optimize unction repeatedly, but I get the error message "too many input argument" on simulation. Please I need assistance, the following is the code:
function L = LMI(ch)
    persistent gamma_square alpha beta C P
    %%%%%%%  Strip Stability Domain Parameters for unit circle %%%%%%%
    alpha = -0.75;               % minimum real part of eigenvalues
    beta = 0.75;                  % maximum real part of eigenvalues 
    eAT = [];
    for i = 1:8:8*8
        eAT = [eAT,ch(i:i+7)];
    end 
    % A_aug;
    Bd = [];
    for i = 64+1:8:65+15
        Bd = [Bd,ch(i:i+7)];
    end
    C=[eye(2) zeros(2) zeros(2) zeros(2)];
    %%%%%%% LMI Design Using YALMIP %%%%%%%%
    yalmip('clear')
    % %%% LMI Variables 
    ops = sdpsettings('solver','sedumi');   % Choice of solver 
    X = sdpvar(8,8);                % 8x8 fully parameterized square matrix  
    Y = sdpvar(2,8);                % 2x8 matrix   
    Z = sdpvar(2,2);                % 2x2 symmetrical matrix 
    gamma_square = sdpvar(1,1);     % Upper bound of H2 norm to minimize

    % Define your optimization problem with the time-varying constraints
    constraints = [eAT'*X + X*eAT + C'*Y + Y'*C + 2*alpha*X <= 0,
                   eAT'*X + X*eAT + C'*Y + Y'*C + 2*beta*X >= 0,
                   eAT'*X + X*eAT + C'*Y + Y'*C + C'*C <= 0,
                   [-Z, Bd'*X; X*Bd, -X] <= 0,
                   trace(Z) <= gamma_square];
    % Define the objective function
    objective = gamma_square; % Objective function
    % Define an optimizer object which solves the problem for a particular
    % initial state and reference
    if isempty(P)
        P = optimizer(constraints,objective,ops,[],X,Y);
    end
%     % Call the optimizer object with the current variables
    P(X,Y);
    % Extract the observer gain from the optimizer object
    L = -value(X)\value(Y)';

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @Vfriday ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.
